### PR TITLE
border: fix "keep only one if both strong and weak symbol exist"

### PR DESCRIPTION
**Pull request by ampresent**
If there are a strong and a weak one, global_fn_dict will be overwritten because they're both public. Then get_func_decl_strs mistooks one of them as local symbol and fail to assert.

This bugfix takes the advantage of the fact: strong symbols are unique. (Although we hope weak symbols are unique too, but GCC seems to allow that)

So now "global_fn_dict" is actually "global_strong_fn_dict". And now we can divided all functions into 1) strong global 2) weak global 3) local. We use func_class to store all weak symbols. Then if a symbol is neither 1) nor 2), we know it's a local symbol.

Fixes: 89e09ed133e7 ("border: keep only one if both strong and weak symbol exist")
Signed-off-by: Yihao Wu <wuyihao@linux.alibaba.com>

### CODE REVIEW

**Comment by ampresent** [comment]
PR not ready yet

File: boundary/analyze.py [hunk 1]
```diff
@@ -328,12 +329,26 @@ class dotdict(dict):
...
+            if fn.weak:
+                """arch_kexec_apply_relocations_add was defined as weak
+                symbol 2 times in the kernel, and this bug is fixed in
+                "kexec_file: Allow archs to handle special regions while
+                locating memory hole". We can only workadround it."""
+                if fn.name != 'arch_kexec_apply_relocations_add':
+                    assert fn.name not in global_fn_dict, \
+                        "Found multiple weak symbol %s" % fn.name
+                global_fn_dict[fn.name] = fn.file
+
```
Comment (chenshanpei) [inline boundary/analyze.py:351]: Can we put this logic to the next "for meta in metas" loop ?  Adding extra pass looks bloated to me.

Comment (ampresent) [reply boundary/analyze.py:351]: @chenshanpei I agree. And according to testing report from @dtcccc , this PR is very unstable. So it needs to be re-designed.

**Comment by ampresent** [comment]
Please ignore all comments above as the implementation is completely refactored

**Comment by Dengerwei** [comment]
LGTM

**Pull request merged**